### ISSUE TITLE
badgerds fails to initialize (after a kill?)

### ISSUE THREAD

**Issue by Voker57** [ipfs/go-ipfs#4363]
#### Version information:

go-ipfs version: 0.4.12-rc1-68a2d4d42
Repo version: 6
System version: amd64/linux
Golang version: go1.8.1

#### Description:
So I was debugging a failing `ipfs add`, killing daemon a few times, and now it won't launch:

```
Initializing daemon...
2017/11/02 03:16:08 Assert failed
gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger/y.AssertTrue
	/home/voker57/go/src/gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger/y/error.go:55
gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger.replayFunction.func2
	/home/voker57/go/src/gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger/db.go:159
gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger.(*valueLog).iterate
	/home/voker57/go/src/gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger/value.go:247
gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger.(*valueLog).Replay
	/home/voker57/go/src/gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger/value.go:620
gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger.Open
	/home/voker57/go/src/gx/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger/db.go:285
gx/ipfs/QmNm2bfBCtEzmpxVzaW2FZGzWx2KWHprfd27aWoXptrAGa/go-ds-badger.NewDatastore
	/home/voker57/go/src/gx/ipfs/QmNm2bfBCtEzmpxVzaW2FZGzWx2KWHprfd27aWoXptrAGa/go-ds-badger/datastore.go:35
github.com/ipfs/go-ipfs/repo/fsrepo.(*badgerdsDatastoreConfig).Create
	/home/voker57/go/src/github.com/ipfs/go-ipfs/repo/fsrepo/datastores.go:393
github.com/ipfs/go-ipfs/repo/fsrepo.measureDatastoreConfig.Create
	/home/voker57/go/src/github.com/ipfs/go-ipfs/repo/fsrepo/datastores.go:335
github.com/ipfs/go-ipfs/repo/fsrepo.(*measureDatastoreConfig).Create
	<autogenerated>:5
github.com/ipfs/go-ipfs/repo/fsrepo.(*FSRepo).openDatastore
	/home/voker57/go/src/github.com/ipfs/go-ipfs/repo/fsrepo/fsrepo.go:407
github.com/ipfs/go-ipfs/repo/fsrepo.open
	/home/voker57/go/src/github.com/ipfs/go-ipfs/repo/fsrepo/fsrepo.go:166
github.com/ipfs/go-ipfs/repo/fsrepo.Open.func1
	/home/voker57/go/src/github.com/ipfs/go-ipfs/repo/fsrepo/fsrepo.go:110
github.com/ipfs/go-ipfs/repo.(*OnlyOne).Open
	/home/voker57/go/src/github.com/ipfs/go-ipfs/repo/onlyone.go:35
github.com/ipfs/go-ipfs/repo/fsrepo.Open
	/home/voker57/go/src/github.com/ipfs/go-ipfs/repo/fsrepo/fsrepo.go:112
main.daemonFunc
	/home/voker57/go/src/github.com/ipfs/go-ipfs/cmd/ipfs/daemon.go:237
github.com/ipfs/go-ipfs/commands.(*Command).Call
	/home/voker57/go/src/github.com/ipfs/go-ipfs/commands/command.go:116
main.callCommand
	/home/voker57/go/src/github.com/ipfs/go-ipfs/cmd/ipfs/main.go:355
main.(*cmdInvocation).Run
	/home/voker57/go/src/github.com/ipfs/go-ipfs/cmd/ipfs/main.go:192
main.mainRet
	/home/voker57/go/src/github.com/ipfs/go-ipfs/cmd/ipfs/main.go:157
main.main
	/home/voker57/go/src/github.com/ipfs/go-ipfs/cmd/ipfs/main.go:64
runtime.main
	/usr/lib/go-1.8/src/runtime/proc.go:185
runtime.goexit
	/usr/lib/go-1.8/src/runtime/asm_amd64.s:2197
```

**Comment by Stebalien** [comment]
Awesome (well, not for you but it's nice to see bugs like this before we make features non-experimental 😛). If possible, could you *back up* your repo and upload it somewhere (or add it to a new IPFS repo and post an IPFS link to it).

**Comment by Stebalien** [comment]
cc @deepakjois

**Comment by deepakjois** [comment]
Yes, a .tar.gz backup of the Badger datastore would help.

This error seems to be happening at initialization, when Badger tries to make sure that there were no crashes the last time the DB was opened, and perform crash recovery if needed.

**Comment by lgierth** [comment]
off topic:

@deepakjois Hi fellow ex-soundclouder! Great to meet you again :) https://soundcloud.com/l_rs/deepak-getting-shit-done-at

**Comment by Stebalien** [comment]
@deepakjois you can download an exact copy of the code he was using at `https://ipfs.io/ipfs/Qmao4ZjrenhCnyy3vmZ1q4rjvbo7BZtKgWpYujtpbXwJu8/badger` (one of the nice things about gx).

**Comment by Voker57** [comment]
So I tried to reproduce the issue but failed. I'd upload the repo, but it's 242gb and probably contains materials forbidden to distribute on ipfs community.

**Comment by Stebalien** [comment]
@deepakjois is there any subset of the database he can upload that may give you a clue? For example, an index/log?

**Comment by leerspace** [comment]
@Voker57 I know you haven't had success reproducing it, but what were you doing when it originally happened? Is it just killing the daemon while an `ipfs add` is in progress? What about the `ipfs add` in your original post was failing?

I'd be willing to spend some time trying to reproduce this and will do so with a repo that can be shared.

**Comment by Voker57** [comment]
@leerspace I think it was generic error like "connection closed", logs of daemon when adding are unfortunately lost. I was adding ~350GB of 10-100mb files. I also might have killed daemon with SIGINT in process.

**Comment by Voker57** [comment]
Oh, and another thing -- I was having some problems with open files limit. I included limits of open files in my reproduction attempts but had no luck.

**Comment by leerspace** [comment]
TL;DR: I haven't been able to reproduce it either.

I wrote a script that starts up and ipfs daemon, starts adding a random file to ipfs (I also tried with a folder of 10000 small files), then kills the daemon ungracefully. I ran it through 1000 iterations and couldn't manage to reproduce it, so I'm not sure if there's some other factor (maybe timing of killing the daemon?). Since connection closing has been implemented I haven't had any issues with open files limits, so maybe the number of connections is a factor.

I did manage to get another error (though I haven't been able to reproduce it), but it didn't prevent me from starting up the daemon again.

**Comment by schomatis** [comment]
This is a great example of a sudden stop (fast daemon shutdown, process kill, system crash) that may not compromise big data loss but would still render the repository unusable for the user (because of a timestamp inconsistency during the interrupted commit), where I'm not sure the user has any option (at the Badger level) to forgo the replay of the truncated data (accepting data loss in exchange for retaining the consistent part of the database).

@leerspace Could you share the script you were using so as to have it as a base for more elaborate tests?

@Voker57 By any chance do you still have the corrupted repo? Maybe just uploading the `MANIFEST` file and the last couple of `sst` and `vlog` files might (not sure) be enough to reproduce the error.

**Comment by leerspace** [comment]
@schomatis I don't have the script anymore, unfortunately. I deleted it after I was unable to reproduce the issue with it.

**Comment by schomatis** [comment]
Looking at the trace there seems to be a mismatch in transaction timestamps while replicating the DB, but without the actual DB there doesn't seem much we can do here, especially considering the Badger version used is rather old. @Voker57 Could you find any remains of the failing DB? Otherwise I'll be closing this issue.

**Comment by Voker57** [comment]
@schomatis sorry, forgot about that first time you highlighted me. I still have the DB, but, as I said earlier, it likely contains content forbidden on IPFS community so I can't share it. If there are some files not containing actual data, I could post them, I guess.

**Comment by schomatis** [comment]
@Voker57 Yes, the `MANIFEST` and the `.sst` files which contain only the indexes of the DB (the actual data is stored in the `.vlog` files). It's not ideal but at least it will give me a chance to give it a try before closing this issue.

Another thing you could try that might also help is running the `badger info --dir <badgerds dir path>` command (or `badger_info` depending on you Badger version) and posting the output here to see if there is any relevant information.

**Comment by Voker57** [comment]
Requested files: https://ipfs.io/ipfs/Qmc4CNGWsiUEaNBQsHriEZuVPiSNRuyk5u4bdXNWqcv7xT

$ badger info --dir .crashed.ipfs/badgerds/
Error: manifest has unsupported version: 3 (we support 4)

**Comment by leerspace** [comment]
I've just encountered the error in this issue with v0.4.16. At some point something happened such that I received this error when trying to start the daemon: `Error: Unable to replay value log: "C:\\Users\\user\\.ipfs\\badgerds\\000088.vlog": Data corruption detected. Value log truncate required to run DB. This would result in data loss`.

I then ran `badger --vlog-dir badgerds/ --dir badgerds/ backup -t` as described in #5213 and it completed, but now this is what I get when attempting to start the daemon. I'm not sure if it's the same error as what's in the original issue (it looks similar to me), so let me know if this should be a separate issue.

```
>ipfs daemon
Initializing daemon...
2018/07/23 12:36:25 Assert failed
gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger/y.AssertTrue
        /build/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger/y/error.go:55
gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger.replayFunction.func2
        /build/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger/db.go:149
gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger.(*valueLog).iterate
        /build/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger/value.go:305
gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger.(*valueLog).Replay
        /build/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger/value.go:690
gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger.Open
        /build/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmeAEa8FDWAmZJTL6YcM1oEndZ4MyhCr5rTsjYZQui1x1L/badger/db.go:300
gx/ipfs/QmaovxzXCm7NMURpC9U5M6QCyHcumPDYAFSq5Rkr8M8CW1/go-ds-badger.NewDatastore
        /build/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmaovxzXCm7NMURpC9U5M6QCyHcumPDYAFSq5Rkr8M8CW1/go-ds-badger/datastore.go:51
github.com/ipfs/go-ipfs/repo/fsrepo.(*badgerdsDatastoreConfig).Create
        /build/distributions/dists/go-ipfs/gopath/src/github.com/ipfs/go-ipfs/repo/fsrepo/datastores.go:411
github.com/ipfs/go-ipfs/repo/fsrepo.measureDatastoreConfig.Create
        /build/distributions/dists/go-ipfs/gopath/src/github.com/ipfs/go-ipfs/repo/fsrepo/datastores.go:334
github.com/ipfs/go-ipfs/repo/fsrepo.(*measureDatastoreConfig).Create
        <autogenerated>:1
github.com/ipfs/go-ipfs/repo/fsrepo.(*FSRepo).openDatastore
        /build/distributions/dists/go-ipfs/gopath/src/github.com/ipfs/go-ipfs/repo/fsrepo/fsrepo.go:422
github.com/ipfs/go-ipfs/repo/fsrepo.open
        /build/distributions/dists/go-ipfs/gopath/src/github.com/ipfs/go-ipfs/repo/fsrepo/fsrepo.go:170
github.com/ipfs/go-ipfs/repo/fsrepo.Open.func1
        /build/distributions/dists/go-ipfs/gopath/src/github.com/ipfs/go-ipfs/repo/fsrepo/fsrepo.go:114
github.com/ipfs/go-ipfs/repo.(*OnlyOne).Open
        /build/distributions/dists/go-ipfs/gopath/src/github.com/ipfs/go-ipfs/repo/onlyone.go:35
github.com/ipfs/go-ipfs/repo/fsrepo.Open
        /build/distributions/dists/go-ipfs/gopath/src/github.com/ipfs/go-ipfs/repo/fsrepo/fsrepo.go:116
main.daemonFunc
        github.com/ipfs/go-ipfs/cmd/ipfs/daemon.go:238
gx/ipfs/QmNueRyPRQiV7PUEpnP4GgGLuK1rKQLaRW7sfPvUetYig1/go-ipfs-cmds.(*executor).Execute
        /build/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmNueRyPRQiV7PUEpnP4GgGLuK1rKQLaRW7sfPvUetYig1/go-ipfs-cmds/executor.go:108
gx/ipfs/QmNueRyPRQiV7PUEpnP4GgGLuK1rKQLaRW7sfPvUetYig1/go-ipfs-cmds/cli.Run.func4
        /build/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmNueRyPRQiV7PUEpnP4GgGLuK1rKQLaRW7sfPvUetYig1/go-ipfs-cmds/cli/run.go:146
runtime.goexit
        /home/why/go/src/runtime/asm_amd64.s:2361
```

I'd be happy to share any repo files if would help.

**Comment by schomatis** [comment]
Hey @leerspace, thanks for reporting this, would you mind opening a separate issue (and point to this one if you want to avoid the trouble of repeating the dump) about it? This is certainly the same assert that's failing as the original one, but the cause of yours seems to be a repo migration (that even terminated successfully) which is very different (I'm assuming) from the original crash/kill scenario, and also because Badger has change a lot in these past 9 months.

(Also, but let's discuss this further in the separate issue, could you repeat the migration to see if you have the same problem?)

**Comment by magik6k** [comment]
There were many fixes and changes in Badger since this issue was opened and I haven't seen this on any of my nodes (some of which are restarted / killed frequently).

Feel free to reopen / comment if you still see this
